data-analytics/cloud-composer-examples/composer_http_post_example/spark_avg_speed.py

```diff
@@ -0,0 +1,144 @@
+                    with an average_speed field. (This argument object gets modified in place by
+                    this method).
+        """
+        # Check if store_and_fwd_flag is a legal value 'Y' or 'N'
+        # (there is some data quality issue in the public table chosen for this example).
+        if record.get('store_and_fwd_flag') and record.get('store_and_fwd_flag') not in 'YN':
+            record['store_and_fwd_flag'] = None
+
+        # Check that fields necessary for calculation are not empty for this record.
+        if (record['pickup_datetime'] and record['dropoff_datetime']
+                and record['trip_distance'] > 0):
+            # Parse strings output by BigQuery to create datetime objects
+            time_0 = datetime.datetime.strptime(record['pickup_datetime'],
+                                                '%Y-%m-%d %H:%M:%S UTC')
+            time_1 = datetime.datetime.strptime(record['dropoff_datetime'],
+                                                '%Y-%m-%d %H:%M:%S UTC')
+            # Calculate a time_delta object between the pickup and drop off times.
+            elapsed = time_1 - time_0
+            if elapsed > datetime.timedelta(0):  # Only calculate if drop off after pick up.
+                # Calculate speed in miles per hour.
+                record['average_speed'] = (3600.0 * record['trip_distance']) / \
+                                           elapsed.total_seconds()
+            else:  # Speed is either negative or undefined.
+                record['average_speed'] = None
+        elif record['trip_distance'] == 0.0:
+            record['average_speed'] = 0.0
+        else:  # One of the fields required for calculation is None.
+            record['average_speed'] = None
+
+        # Writes a csv instead of json
+        csv_record = self.dict_to_csv(record)
+        return csv_record
+
+
+def main(sc, gcs_path_raw, gcs_path_transformed):
+    ase = AverageSpeedEnhancer()
+    # Create an AverageSpeedEnhancer instance which contains our business logic
+    file_strings_rdd = sc.textFile(gcs_path_raw)
+    # Apply the speed enhancement logic defined in the AverageSpeedEnhancer class
+    # Read the newline delimited json into dicts (note that this is automatically applied per line).
+    records_rdd = file_strings_rdd.map(lambda record_string: json.loads(record_string))
+    # Enhance dicts with an average speed field.
+    transformed_records_rdd = records_rdd.map(ase.enhance_with_avg_speed)
+    # Save output to the timestamped output directory.
+    transformed_records_rdd.saveAsTextFile(gcs_path_transformed)
+
+
+if __name__ == "__main__":
+
+    # Configure Spark
+    spark_conf = SparkConf()
+    spark_conf.setAppName('AverageSpeedEnhancement')
+    spark_context = SparkContext(conf=spark_conf)
+
+    parser = argparse.ArgumentParser()
+
+    # This is the path to the existing files in GCS that are the candidates for enhancement.
```

Comment:
This is redundant with the help text, delete.

Reply:
Deleted